engine/crates/engine-v2/schema/src/input_value/raw.rs

```diff
@@ -0,0 +1,522 @@
+use crate::{
+    EnumValueId, IdRange, InputValueDefinitionId, RawInputKeyValueId, RawInputObjectFieldValueId, RawInputValueId,
+    SchemaInputValueId,
+};
+
+/// Holds input values for the Schema and for an Operation during execution.
+/// It's generic over 'Str', the string representation:
+/// - StringId for the Schema (default input values and directive arguments)
+/// - Box<str> for the Operation
+///
+/// This allow us to share the code for:
+/// - Display: used to print default values in introspection and add input values in subgraph query strings.
+/// - Serialize: used to serialize variables/arguments.
+/// - Deserializer: used by resolvers to deserialize arguments into a specific struct.
+///
+/// Data is stored in flat arrays for hopefully faster serialization and as as bonus smaller InputValue
+/// footprint on x64 architectures: u64 + usize (padding) rather than 3 * usize (Box<[]> + padding).
+#[derive(Clone)]
+pub struct RawInputValues<Str> {
+    /// Inidividual input values and list values
+    pub(super) values: Vec<RawInputValue<Str>>,
+    /// InputObject's fields
+    pub(super) input_fields: Vec<(InputValueDefinitionId, RawInputValue<Str>)>,
+    /// Object's fields (for JSON)
+    pub(super) key_values: Vec<(Str, RawInputValue<Str>)>,
+}
+
+impl<Str> Default for RawInputValues<Str> {
+    fn default() -> Self {
+        Self {
+            // Reserve the first slot for undefined values.
+            // This allows us in the engine to create from scratch a PlanInputValue that is
+            // undefined. Used when requesting a field argument that wasn't provided.
+            values: vec![RawInputValue::Undefined],
+            input_fields: Default::default(),
+            key_values: Default::default(),
+        }
+    }
+}
+
+/// Represents any possible input value for field arguments, operation variables and directive
+/// arguments.
+/// Used for storage and as input format for operation input values to support default values,
+/// variables, etc.
+#[derive(Debug, Clone)]
+pub enum RawInputValue<Str> {
+    Null,
+    String(Str),
+    EnumValue(EnumValueId),
+    Int(i32),
+    BigInt(i64),
+    Float(f64),
+    Boolean(bool),
+    InputObject(IdRange<RawInputObjectFieldValueId<Str>>),
+    List(IdRange<RawInputValueId<Str>>),
+
+    /// for JSON
+    Map(IdRange<RawInputKeyValueId<Str>>),
+    U64(u64),
+
+    /// Directive arguments have unknown enum values. Likely doesn't make sense, but not sure.
+    UnknownEnumValue(Str),
+
+    /// Primary purpose is for operation to reference the value of a variable which is initialized
+    /// to the default or an arbitrary value first. It might also be used later to share the same
+    /// InputValue at different places.
+    Ref(RawInputValueId<Str>),
```

Comment:
I don't get this at all.  If it's just a variable: ok, I probably get it (though I don't think it'd hurt to have the word in the variant name somewhere).

But I assume from the comment it's not: but for what purpose do we want to "use later to share the same InputValue at different places"?  Why do we only store variables "which are initialized to the default or an arbitrary value first"?  Do we put variables that aren't initialised somewhere else?  Am I just misunderstanding entirely?

Reply:
well ok, not that clear. with a proper name it would be `VariablePlaceholderRef` which felt a bit weird to put inside `engine-v2-chema`. But that might be my whole problem, trying to avoid duplicating the serde/deser/etc. implementation by having a more complex struct playing the role of both `engine::Value` and `engine::ConstValue`.

And I thought somehow it would be smart to update input values we parse from the query with the variables to avoid more back and forth, but it's just more complex in the end. I should probably keep the `ConstValue` vs `Value` separation and just duplicate code